common/src/chain/transaction/signed_transaction.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{signature::inputsig::InputWitness, Transaction};
+use crate::primitives::{
+    id::{self, Idable, WithId},
+    Id,
+};
+use serialization::{Decode, Encode};
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct SignedTransaction {
+    transaction: Transaction,
+    signatures: Vec<InputWitness>,
+}
+
+impl SignedTransaction {
+    // TODO(PR) return Result instead of just Self, and check that witness size is equal to Transaction inputs size
+    pub fn new(transaction: Transaction, signatures: Vec<InputWitness>) -> Self {
+        Self {
+            transaction,
+            signatures,
+        }
+    }
+
+    pub fn transaction(&self) -> &Transaction {
+        &self.transaction
+    }
+
+    pub fn signatures(&self) -> &[InputWitness] {
+        self.signatures.as_ref()
+    }
+
+    /// provides the hash of a transaction including the witness (malleable)
+    pub fn serialized_hash(&self) -> Id<Transaction> {
+        Id::new(id::hash_encoded(self))
+    }
+}
+
+impl Idable for SignedTransaction {
+    type Tag = Transaction;
```

Comment:
The assumption is that if `Id<A> == Id<B>`, then `A == B`. The problem with `Id<SignedTransaction>` is that it's malleable, so it's not unique. You can very well have `Id<A> == Id<B>` false, but still have A == B true, when you ignore the signature. This all drills down to why we invented Idable in the first place. A signed transaction isn't Idable, because it can't have a unique Id.

Reply:
> The assumption is that if `Id<A> == Id<B>`, then `A == B`. The problem with `Id<SignedTransaction>` is that it's malleable, so it's not unique. You can very well have `Id<A> == Id<B>` false, but still have A == B true, when you ignore the signature. This all drills down to why we invented Idable in the first place. A signed transaction isn't Idable, because it can't have a unique Id.

The idea is that `Id<SignedTransaction>` is an ID for particular pair of (Transaction, Signatures) because the signatures would be included in the ID hash. That's what would differentiate `Id<SignedTransaction>` from `Id<Transaction>`. Changing the signature would change the ID, so the ID is unique if you consider the signature to be part of the object you're taking ID of.

Reply:
I understand what you mean when you look at it within the data perspective, but it's a bad idea in the big picture. The whole reason we did the Idable thing is uniqueness, and this will kill that. Think of it like this: What I want is, if `Id<A> == Id<B>`, then `A == B` even if the signatures are the same. What's the point of making them idable?

It's like me getting a new id if I change my outfit... well... some people won't like that :D

This is what serialized_hash() is for. It covers malleability, for things that are not Idable.

Reply:
I'm OK with removing Idable from `SignedTransaction` to reduce confusion. I also told Roy to change serialized_hash to return H256. Does that sound better?

Reply:
> I'm OK with removing Idable from `SignedTransaction` to reduce confusion. I also told Roy to change serialized_hash to return H256. Does that sound better?

Not sure, `H256` is not a very precise type as it does not include what it represents (i.e. hash of a signed transaction). It's fine with me for now but we should think about something better.

Reply:
Agreed. H256 just because I don't have a solution yet to this.